Adding dinov3 collection of vit models

coder0143 opened the pull request:
Resolves #98 

<!--- Please check [issues](https://github.com/jax-ml/bonsai/issues) for any pending model implementations. Consider opening issue if none exists. -->

**Reference**
* Huggingface implementation: https://github.com/huggingface/transformers/blob/main/src/transformers/models/dinov3_vit/modular_dinov3_vit.py

**Checklist**
<!--- Please make sure all checkboxes are ticked before submitting this PR for review. -->

- [x] I have read the **[Contribution Guidelines](https://github.com/jax-ml/bonsai/blob/main/CONTRIBUTING.md#contributing-a-model)** and used [pre-commit hooks](https://github.com/jax-ml/bonsai/blob/main/CONTRIBUTING.md#linting-and-type-checking) to format this commit.
- [x] I have added all the necessary **unit tests** for my change. (`run_model.py` for model usage, `test_outputs.py` and/or `model_validation_colab.ipynb` for quality).
- [x] **(If using an LLM)** I have carefully reviewed and removed all **superfluous comments** or unneeded, commented-out code. Only necessary and functional code remains.
- [x] I have signed the **[Contributor License Agreement (CLA)](https://cla.developers.google.com/about)**.

coder0143 commented:
Can you please review my PR @chapman20j

chapman20j commented on bonsai/models/dinov3/params.py:
```diff
@@ -0,0 +1,165 @@
+    return m
+
+
+def load_model_config(model_path: str) -> DINOv3ViTConfig:
```
Can we put our own model configs in the `modeling.py` file? This would be consistent with other implementations on Bonsai. It could also be used to enforce some assumptions about the code (e.g. using `tuple[int]` for the patch size).

chapman20j commented on bonsai/models/dinov3/modeling.py:
```diff
@@ -0,0 +1,234 @@
+from typing import Tuple
+from transformers import DINOv3ViTConfig
+
+def make_2tuple(x):
```
Similar to my other comment (https://github.com/jax-ml/bonsai/pull/104/changes#r2621001569), we can specify our own configs to avoid this check.

jenriver commented:
Hi, could you ensure that the tests above are passing? i.e.

1) Please ensure you have run `pre-commit run --all-files` as in [contribution guidelines](https://github.com/jax-ml/bonsai/blob/main/CONTRIBUTING.md#bonsai-pull-request-checklist).
2) The CI is currently failing with a `GatedRepoError` because the dinov3 checkpoint is restricted. Could you update the test to use randomly initialized weights instead? (Note: Please ensure the JAX and PyTorch models are initialized with the same random weights so the parity assertions still pass. Creating a random PyTorch model and converting it to JAX within setUp usually works best.)

coder0143 commented:
I have made the following changes:
* Checked and updated files based on ruff formatting. 
* Removed run_model.py file, instead colab notebook can be used.
* Updated test_outputs.py file to use a randomly initialized model and run tests based on that, using the vit_b16 model

jenriver commented:
Hi, we're still seeing pre-commit failures as above -- could you ensure you have run pre-commit hooks?

i.e.
`pre-commit run --all-files` as in [contribution guidelines](https://github.com/jax-ml/bonsai/blob/main/CONTRIBUTING.md#bonsai-pull-request-checklist).

jenriver commented on bonsai/models/dinov3/tests/test_outputs.py:
```diff
@@ -0,0 +1,125 @@
+    def setUp(self):
+        super().setUp()
+        raw_path = "~/.cache/huggingface/dinov3_vitb16"
+        self.save_dir = os.path.expanduser(raw_path)
```
Could you not use local directory paths?

i.e. Something like 

`self.save_dir = snapshot_download(...)`
